Lib/concurrent/futures/_base.py

```diff
@@ -520,7 +521,7 @@ def submit(self, fn, *args, **kwargs):
         """
         raise NotImplementedError()
 
-    def map(self, fn, *iterables, timeout=None, chunksize=1):
+    def map(self, fn, *iterables, timeout=None, chunksize=1, prefetch=None):
         """Returns an iterator equivalent to map(fn, iter).
```

Comment:
This change will also affect `ProcessPoolExecutor`. However, for `ProcessPoolExecutor`, `chunksize` argument already provides a similar functionality as `prefetch`: while it's not documented, the current implementation will grab precisely `chunksize` items from the input and won't ask for more until the first batch is fully processed. Therefore, if `prefetch + self._max_workers >= chunksize`, your PR will not change the current behavior. If `prefetch + self._max_workers < chunksize`, your PR will have an undesired effect of reducing the effective number of items sent to the remote process, defeating the optimization produced by the user-selected value `chunksize`.

Perhaps you'd want to restrict your PR solely to `ThreadPoolExecutor`?

Reply:
The documentation for `chunksize` describes the resulting chunks as tasks (so a 100 long input with `chunksize=10` is doing 10 tasks). Similarly, I documented the `prefetch` argument as a task count, not an input count, so dividing it by `chunksize` would be incorrect.

Changing the documentation of `prefetch` would be an option, but I think having `prefetch` defined in terms of tasks is more intuitive in any event; the default `prefetch` count being `_max_workers` is intended to ensure that, in the typical case where individual tasks are usually take approximately the same order of magnitude (and that time is significantly longer than the overhead of dispatching the task and retrieving results), enough tasks on deck that, even if all outstanding tasks finish simultaneously, or it takes nearly as long to process each result as it does to produce an input, or the next task to be waited on takes longer than the subsequent tasks, the workers don't go idle while the caller is processing the results. If you divide the `prefetch` count by `chunksize`, then increasing the `chunksize` without adjusting `prefetch` in tandem means your buffer of tasks drops below the number of workers. For example, on a four worker pool, if call 1 takes 10 seconds (for whatever reason), and all other calls take 6 seconds, with default arguments, calls 1-8 are queued up initially, with 1-4 running immediately; when calls 2, 3 and 4 finish, calls 5, 6 and 7 begin running immediately, so the pool stays fully occupied. If increasing the `chunksize` reduced the `prefetch` value, then passing `chunksize=2` would mean that you submitted a task chunking calls 1 & 2, 3 & 4, 5 & 6, 7 & 8 (all of which are dispatched immediately) with only 9 & 10 and 11 & 12 on deck. Now, when 3 & 4, 5 & 6, 7 & 8 complete (after ~12 seconds), two of the three freed up workers immediately have work to do (9&10 and 11&12), but the final worker is idle for four more seconds, at which point 1&2 finishes, the results are retrieved and 13&14 is pulled and dispatched. Except the worker that had been doing 1&2 is now idle, and if the work done to process the results is non-trivial (say, takes a second each), we now have to wait until both 1 and 2 are processed before 3 & 4 are requested, during which time we've spent two more seconds with an idle worker.

I'm open to criticisms of the default `prefetch` count, but the default should not be altered by changing the `chunksize`; if the user wants to submit larger chunks, it has no impact on how much of a buffer we want to maintain to ensure workers remain occupied.

As for "avoid increasing it by another `_max_workers` in `map`", if you're referring to the `islice` end index being `self._max_workers + prefetch`, that's not actually increasing the `prefetch`; that's the "fetch" count. `prefetch` is "tasks submitted in excess of the number of workers", but we want to pull at least enough tasks to occupy the workers entirely from the beginning, with `prefetch` being tasks that we `submit` knowing that (barring out-of-order task completion) they're unlikely to be processed immediately.